### ISSUE TITLE
Implement terrain handling for Minecraft 1.13

### ISSUE THREAD

**Issue by Mhowser** [ORelio/Minecraft-Console-Client#599]
https://github.com/cabaletta/baritone

Would it be possible to integrate this into MCC? Might solve /move not working on 1.13.

Edit: I was under the impression it was on 1.13 already, it's 1.12.2, but could be a stepping stone towards 1.13 movement support.

**Comment by Mhowser** [comment]
I forgot that mcc was made in C# 😅 /move does "work" in 1.12.x, it could
certainly use some enhancing though.

**Comment by ORelio** [comment]
Hi,
Currently, MCC implements pathfinding for moving in a world of blocks. It does not necessarily know about some special blocks such as slabs, but the pathfinding mechanism is still pretty generic.
However in 1.13 terrain format changed. MCC does not know how to process data from the server to build its internal representation of the world. As such, it cannot determine a path.
So to fix pathfinding we do not need to reimplement it, we rather need to implement terrain data decoding for Minecraft 1.13.

**Comment by vkorn** [comment]
@ORelio I was able to fix terrain for 1.14, however, I'm using completely re-written approach, which we've discussed in https://github.com/ORelio/Minecraft-Console-Client/pull/689 
To avoid parsing palette files in memory (and carrying extra 2mb in the binary), I made a simple pre-generator (https://github.com/vkorn/Minecraft-Console-Client/tree/various-features/MaterialGenerator) for simplified list of mats https://github.com/vkorn/Minecraft-Console-Client/blob/various-features/MinecraftClient/Protocol/WorldProcessors/BlockProcessors/114Pre5/Material.cs 

Pretty much you can copy and paste this part into current implementation to fix terrain.

**Comment by ORelio** [comment]
Hi @vkorn thanks for your work on this.

I've read how the 1.13+ palette thing works, and I think it's clever since it greatly reduces chunk data size, although not practical for our implementation which tries to be light and needs to have more and more static Mojang assets to work with newer MC versions (this reminds me of translation files for chat in MC 1.6).

The way you simplified it for MCC seems a good approach as it retains enough data about each block for pathfinding, while avoiding to use the full palette file. However, maybe some bots will need that data. For instance, the [WatchLamp](https://github.com/ORelio/Minecraft-Console-Client/blob/master/MinecraftClient/config/sample-script-with-world-access.cs) bot would not work anymore. The same way, I see your [Material.cs](https://github.com/vkorn/Minecraft-Console-Client/blob/various-features/MinecraftClient/Protocol/WorldProcessors/BlockProcessors/114Pre5/Material.cs) has a `Ore` entry, that you might want to use for something else than pathfinding.

For this reason, maybe actually using the palette file would help having all the information a bot would need? The file could be automatically downloaded or extracted from installed game to avoid including large files in the repository or MCC binary. Maybe we could have a reduced functionality mode in which basic block properties would be known for pathinding but not everything, in case the palette file is missing?

I'm thinking of something similar to the existing [Material.cs](https://github.com/ORelio/Minecraft-Console-Client/blob/master/MinecraftClient/Mapping/Material.cs) which would still implement methods such as `IsSolid` for the sake of pathfinding, but would no longer hardcode block IDs. An additional method would give corresponding data for the currently running Minecraft version pallette. This would give default data for pre-1.13 (we might need to keep existing hardcoded values for this).

Still suggesting for now.

**Comment by vkorn** [comment]
Hey @ORelio, you can, that's what generator is doing right now :) It has list of glob-mappings to parse materials [names](https://github.com/vkorn/Minecraft-Console-Client/blob/various-features/MaterialGenerator/114Pre5/Rules.cs#L25). You can add as many materials as you want, but at the end you still have to manually define mapping "ingame name <=> bot material". And matching this real-time is not a good idea, singe regexp/glob matches are heavy, so you'll need to have at least some caching. I think, for making such bots as WatchLamp work, it would be better to check block's minecraft:name == "redstone_lamp" rather that materials enum.

**Comment by ORelio** [comment]
Hi,

For your information, on my side I've adjusted terrain parsing for 1.13, which happens to be very similar to 1.9 terrain parsing that already has a per-chunk palette feature.

All that's missing would be `Block State Id` => `Material` mapping. After thknking about it, maybe we could add missing entries to Material.cs and add mappings for each version. Since `Material.cs` already defines whether a block is solid and so on, that would avoid defining it again for each new version. However, we'll likely have to create a new mapping for each MC version so automating that still seems relevant.

On a side note, since we're not far away, I think completing 1.13 support would be a good idea before adding 1.14 support 😅 (#704)

**Comment by vkorn** [comment]
I ended up loading palette from an archive upon startup and exposing properties, something like this

```
>/inspect -1806 22 1148
21:21:17 MCC: Material: lever
Properties: face: wall facing: north powered: true
```

Since imo exposing all potentially viable properties through fixed enum is close to impossible due to huge amount of combinations.

**Comment by ORelio** [comment]
OK so after much trial and error, here is what I could come up with: https://github.com/ORelio/Minecraft-Console-Client/commit/c04b17cabc07940cf1727e2093566d0ba5b2f6a1

Basically, I tried to add Palette support while creating as few changes as possible in the code to avoid creating bugs and retain similar functionality. Thanks @TheSnoozer and @vkorn for your research that greatly helped in choosing an approach.

The `Material.cs` enum is still there and very similar to its previous form, but a generic `id => Material` dictionary (the Palette) will map block IDs to the corresponding Material entries for the current Minecraft version instead of hardcoding IDs directly inside the enum.

Block states for the same block type are mapped to the same `Material` item, so we're losing some data such as redstone lamp state. Likewise, as old `BlockMeta` is not handled in the new `id => Material` mapping, the pre-1.13 compatibility palette will map some blocks to their default variant: MC 1.12 wool is considered to be white regardless of the actual color, and so on (this does not affect MC 1.13+).

Despite its simple form, the generated 1.13 palette is a 380 KB class, that I hope to be smaller in its compiled form as it only contains a static `int => Material` dictionary. For performance reasons, the block state is currently stored in the same `ushort` field as before (this also saves RAM as we're still far to filling all the 65536 possibilities) and the palette is queried only when the block type needs to be known for movement and pathfinding. When debugging, I could confirm that despite being a simple dictionary, the palette lookup cost was significant if performed for all blocks when loading the chunk, so lazy lookup should be preferred.

For newer Minecraft versions, the palette update steps should be straightforward:
1. Generate `blocks.json` using data reporting on Vanilla Minecraft
2. Generate temporary `MaterialXXX.cs` and `PaletteXXX.cs` using PaletteGenerator
3. Perform a diff with existing versions, add missing entries in `Material.cs` and `MaterialExtensions.cs`
4. If existing state IDs were _not_ randomized by Mojang, add missing entries to `Palette113.cs`
5. If existing state IDs were randomized, add a new palette as `PaletteXXX.cs` into the codebase (sigh)

As usual, bugs are likely present, so please test the latest snapshot and report them!

**TL;DR:** Terrain and movements implemented for 1.13! 🎉
